samza-core/src/main/java/org/apache/samza/runtime/ClusterBasedProcessorLifecycleListener.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.runtime;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.TaskConfig;
+import org.apache.samza.container.SamzaContainerListener;
+import org.apache.samza.util.ThreadUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class ClusterBasedProcessorLifecycleListener implements SamzaContainerListener {
+  private static final Logger log = LoggerFactory.getLogger(ClusterBasedProcessorLifecycleListener.class);
+  private final Runnable onSignal;
+  private Thread shutdownHookThread = null;
+  private CountDownLatch shutdownLatch = new CountDownLatch(1);
+  private TaskConfig taskConfig;
+  private ProcessorLifecycleListener processorLifecycleListener;
+  private volatile Throwable containerException = null;
+
+  ClusterBasedProcessorLifecycleListener(Config config, ProcessorLifecycleListener processorLifecycleListener,
+      Runnable onSignal) {
```

Comment:
This is always the "shutdown container" method, right? Could you please name it to be a little more specific to that? `onSignal` seems a bit vague.

Reply:
Although we currently use this to only call `container.shutdown()`, the callback can be used to execute anything else. The naming here is based on the fact that the callback is invoked in the signal-handler/shutdown hook.
Do you think `shutdownHookCallback` is a better name ?